pkg/reconciler/apis/apibinding/apibinding_reconcile.go

```diff
@@ -237,209 +116,300 @@ func (r *workspaceAPIExportReferenceReconciler) reconcile(ctx context.Context, a
+			// Existing CRD flow
+			if !apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Established) || apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Terminating) {
 				needToWaitForRequeue = true
-			} else {
-				if !apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Established) || apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Terminating) {
-					needToWaitForRequeue = true
-				}
 			}
 		}
 
+		// Merge any current storage versions with new ones
+		storageVersions := sets.NewString()
+		if existingCRD != nil {
+			storageVersions.Insert(existingCRD.Status.StoredVersions...)
+		}
+
+		for _, b := range apiBinding.Status.BoundResources {
+			if b.Group == schema.Spec.Group && b.Resource == schema.Spec.Names.Plural {
+				storageVersions.Insert(b.StorageVersions...)
+				break
+			}
+		}
+
+		sortedStorageVersions := storageVersions.List()
+		sort.Strings(sortedStorageVersions)
+
 		boundResources = append(boundResources, apisv1alpha1.BoundAPIResource{
 			Group:    schema.Spec.Group,
 			Resource: schema.Spec.Names.Plural,
 			Schema: apisv1alpha1.BoundAPIResourceSchema{
 				Name: schema.Name,
 				UID:  string(schema.UID),
 			},
-			StorageVersions: crd.Status.StoredVersions,
+			StorageVersions: sortedStorageVersions,
 		})
 	}
 
 	conditions.MarkTrue(apiBinding, apisv1alpha1.APIExportValid)
 
-	if needToWaitForRequeue {
-		return reconcileStatusStop, nil
-	}
-
-	conditions.MarkTrue(apiBinding, apisv1alpha1.CRDReady)
-	apiBinding.Status.Initializers = []string{}
-	apiBinding.Status.Phase = apisv1alpha1.APIBindingPhaseBound
 	apiBinding.Status.BoundAPIExport = &apiBinding.Spec.Reference
 	apiBinding.Status.BoundResources = boundResources
 
-	return reconcileStatusContinue, nil
+	if needToWaitForRequeue {
+		conditions.MarkFalse(
+			apiBinding,
+			apisv1alpha1.APIsBound,
+			apisv1alpha1.WaitingForEstablishedReason,
+			conditionsv1alpha1.ConditionSeverityInfo,
+			"Waiting for API(s) to be established",
+		)
+	} else {
+		conditions.MarkTrue(apiBinding, apisv1alpha1.APIsBound)
+		apiBinding.Status.Initializers = []string{}
+		apiBinding.Status.Phase = apisv1alpha1.APIBindingPhaseBound
+	}
+
+	return nil
 }
 
-func (c *controller) reconcile(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
-	reconcilers := []reconciler{
-		&phaseReconciler{
-			getAPIExport:         c.getAPIExport,
-			getAPIResourceSchema: c.getAPIResourceSchema,
-			enqueueAfter:         c.enqueueAfter,
-		},
-		&workspaceAPIExportReferenceReconciler{
-			getAPIExport:         c.getAPIExport,
-			getAPIResourceSchema: c.getAPIResourceSchema,
-			getCRD:               c.getCRD,
-			createCRD:            c.createCRD,
-			updateCRD:            c.updateCRD,
-			deletedCRDTracker:    c.deletedCRDTracker,
-		},
+func (c *controller) reconcileBound(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	apiExportClusterName, err := getAPIExportClusterName(apiBinding)
+	if err != nil {
+		// Should never happen
```

Comment:
can happen if the export gets deleted by the owner, can't it? Also in the future with sharding it might fail here when the APIExport holding shard goes down.